src/Executor/Executor.php

```diff
@@ -222,12 +223,15 @@ private static function getOperationRootType(Schema $schema, OperationDefinition
     private static function executeFieldsSerially(ExecutionContext $exeContext, ObjectType $parentType, $sourceValue, $path, $fields)
     {
         $results = [];
+        $sourceValue = self::completePromiseIfNeeded($sourceValue);
+
         foreach ($fields as $responseName => $fieldASTs) {
             $fieldPath = $path;
             $fieldPath[] = $responseName;
             $result = self::resolveField($exeContext, $parentType, $sourceValue, $fieldASTs, $fieldPath);
 
             if ($result !== self::$UNDEFINED) {
+                $result = self::completePromiseIfNeeded($result);
```

Comment:
Looks like your example passes promise to nested field resolvers. I didn't think about it this way, as I only considered examples where resolvers received actual parent value (resolved promise value, not promise itself).

This is interesting approach. I do have doubts about it, yet I am really interested to see where it leads to. Looking forward to see your concrete examples!